File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int dfs(TreeNode* node, int *max_value) {
        if (!node) return 0;
        int left = max(dfs(node->left, max_value), 0);
        int right = max(dfs(node->right, max_value), 0);
        *max_value = max(*max_value, left + right + node->val);
        return max(left, right) + node->val;
    }

    int maxPathSum(TreeNode* root) {
        if (!root) return 0;
        int max_value = INT_MIN;
        dfs(root, &max_value);
        return max_value;
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int findMaxPath(TreeNode root, int[] maxSum) {
        if (root == null) return 0;
        int l = Math.max(findMaxPath(root.left, maxSum), 0);
        int r = Math.max(findMaxPath(root.right, maxSum), 0);
        int currSum = l + r + root.val;
        maxSum[0] = Math.max(maxSum[0], currSum);
        return Math.max(l, r) + root.val;
    }

    public int maxPathSum(TreeNode root) {
        if (root == null) return 0;
        int[] maxSum = new int[] {Integer.MIN_VALUE};
        findMaxPath(root, maxSum);
        return maxSum[0];
    }
}